bot/mode.py

```diff
@@ -0,0 +1,80 @@
+import logging
+from functools import wraps
+from typing import Callable
+
+from telegram import Update
+from telegram.ext import Updater, CommandHandler, CallbackContext, run_async
+
+from filters import admin_filter
+
+logger = logging.getLogger(__name__)
+
+ON, OFF = True, False
+
+
+class Mode:
+    def __init__(self, mode_name: str, default: bool = True, pin_info_msg: bool = False) -> None:
+        self.name = mode_name
+        self.default = default
+        self.chat_data_key = self._gen_chat_data_key(mode_name)
+        self.pin_info_msg = pin_info_msg
+
+    @staticmethod
+    def _gen_chat_data_key(mode_name: str) -> str:
+        return f"is_{mode_name}_on".lower()
+
+    def _get_mode_state(self, context: CallbackContext):
+        if self.chat_data_key not in context.chat_data:
+            context.chat_data[self.chat_data_key] = self.default
+
+        return context.chat_data[self.chat_data_key]
+
+    def _set_mode(self, state: bool, context: CallbackContext):
+        context.chat_data[self.chat_data_key] = state
+
+    def add(self, func) -> Callable:
+        @wraps(func)
+        def wrapper(upd: Updater, handlers_group: int):
+            dp = upd.dispatcher
+            dp.add_handler(CommandHandler(f"{self.name}_on", self.mode_on, filters=admin_filter), handlers_group)
+            dp.add_handler(CommandHandler(f"{self.name}_off", self.mode_off, filters=admin_filter), handlers_group)
+            func(upd, handlers_group)
+
+        return wrapper
+
+    def handler(self, func) -> Callable:
+        @run_async
+        @wraps(func)
+        def wrapper(update: Update, context: CallbackContext):
+            mode_is_on = self._get_mode_state(context)
+            if mode_is_on:
```

Comment:
Wouldn't it be cleaner solution to do `dp.remove_handler` on mode handlers instead of making handlers to do nothing?

Reply:
Looks like someone has read the documentation!

Reply:
docuu..what? do not recognize the word